tiledb/sm/serialization/array_schema.cc

```diff
@@ -71,6 +73,73 @@ namespace serialization {
 
 #ifdef TILEDB_SERIALIZATION
 
+Status filter_to_capnp(
+    const Filter* filter, capnp::Filter::Builder* filter_builder) {
+  filter_builder->setType(filter_type_str(filter->type()));
+  switch (filter->type()) {
+    case FilterType::FILTER_BIT_WIDTH_REDUCTION: {
+      uint32_t window;
+      RETURN_NOT_OK(
+          filter->get_option(FilterOption::BIT_WIDTH_MAX_WINDOW, &window));
+      auto data = filter_builder->initData();
+      data.setUint32(window);
+      break;
+    }
+    case FilterType::FILTER_POSITIVE_DELTA: {
+      uint32_t window;
+      RETURN_NOT_OK(
+          filter->get_option(FilterOption::POSITIVE_DELTA_MAX_WINDOW, &window));
+      auto data = filter_builder->initData();
+      data.setUint32(window);
+      break;
+    }
+    case FilterType::FILTER_GZIP:
+    case FilterType::FILTER_ZSTD:
+    case FilterType::FILTER_LZ4:
+    case FilterType::FILTER_RLE:
+    case FilterType::FILTER_BZIP2:
+    case FilterType::FILTER_DOUBLE_DELTA:
+    case FilterType::FILTER_DICTIONARY: {
+      int32_t level;
+      RETURN_NOT_OK(
+          filter->get_option(FilterOption::COMPRESSION_LEVEL, &level));
+      auto data = filter_builder->initData();
+      data.setInt32(level);
+      break;
+    }
+    case FilterType::FILTER_SCALE_FLOAT: {
+      FloatScalingFilter::Metadata m;
+      double scale, offset;
+      uint64_t byte_width;
+      RETURN_NOT_OK(
+          filter->get_option(FilterOption::SCALE_FLOAT_BYTEWIDTH, &byte_width));
+      RETURN_NOT_OK(
+          filter->get_option(FilterOption::SCALE_FLOAT_FACTOR, &scale));
+      RETURN_NOT_OK(
+          filter->get_option(FilterOption::SCALE_FLOAT_OFFSET, &offset));
+      m.scale = scale;
+      m.offset = offset;
+      m.byte_width = byte_width;
+      auto data = filter_builder->initData();
+      auto capnpValue = kj::Vector<uint8_t>();
+      capnpValue.addAll(kj::ArrayPtr<uint8_t>(
+          const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(&m)),
+          sizeof(FloatScalingFilter::Metadata)));
+      data.setBytes(capnpValue.asPtr());
```

Comment:
Going the bytes route is dangerous. We need to add a structure for the Metadata, even if this a bit more tedious now, we discourage future attempts of passing complex types as byte streams. Also I would love if we can name it something more specific in the capnp spec, FilterMetadata or FilterConfig, the metadata terminology is already used extensively in that file.